### PR TITLE
ENH Transpose backend primitive 

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This PR adds the transpose backend primitive for all backends (Torch, Tensorflow, Numpy) in 1d scattering, alongside some tests. 

Important to note, this backend primitive only transposes two dimensions. In Tensorflow and Numpy, we transpose the final two dimensions. In Torch, we transpose the second to last and third to last dimensions. 

We should be sure to note the above in some documentation. 

We need to merge pr #807 prior to merging this PR.

### CODE REVIEW

File: tests/scattering1d/test_torch_backend_1d.py [hunk 1]
```diff
@@ -252,3 +252,28 @@ def coefficent(n):
     print(z_2.shape)
     assert not z_2.shape[-1] == 2
     assert torch.allclose(x_r, z_2)
+
+
+@pytest.mark.parametrize("device", devices)
+@pytest.mark.parametrize("backend", backends)
+def test_transpose_1d(device, backend, random_state=42):
+    """
+    Tests the correctness and differentiability of pad_1d
```
Comment (lostanlen) [inline tests/scattering1d/test_torch_backend_1d.py:261]: pad_1d -> transpose_1d

File: tests/scattering1d/test_torch_backend_1d.py [hunk 2]
```diff
@@ -252,3 +252,28 @@ def coefficent(n):
...
+    """
+    shape = (10, 20, 3, 5, 2)
+    shape_T = (10, 20, 5, 3, 2)
+
+    x = torch.ones(shape)
+    x_T = backend.transpose(x)
+    assert tuple(x_T.shape) == shape_T
+
+    x_T_T = backend.transpose(x_T)
+    assert tuple(x_T_T.shape) == shape
```
Comment (lostanlen) [inline tests/scattering1d/test_torch_backend_1d.py:271]: couldn't you be more demanding and assert x_T_T to equal x? not just in terms of shape but in terms of content?

File: tests/scattering1d/test_torch_backend_1d.py [hunk 3]
```diff
@@ -252,3 +252,28 @@ def coefficent(n):
...
+@pytest.mark.parametrize("device", devices)
+@pytest.mark.parametrize("backend", backends)
+def test_transpose_1d(device, backend, random_state=42):
+    """
+    Tests the correctness and differentiability of pad_1d
+    """
+    shape = (10, 20, 3, 5, 2)
+    shape_T = (10, 20, 5, 3, 2)
+
+    x = torch.ones(shape)
```
Comment (lostanlen) [inline tests/scattering1d/test_torch_backend_1d.py:266]: i would prefer x to be [1, 2, 3, 4, 5, 6 ....] reshaped to `shape`
instead of [1, 1, 1, 1, 1, 1, ... ]

File: tests/scattering1d/test_torch_backend_1d.py (hunk 1 above)
Comment (lostanlen) [reply tests/scattering1d/test_torch_backend_1d.py:261]: also are you really testing the differentiability of transpose ? where?

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 4]
```diff
@@ -98,4 +98,10 @@ def ifft(cls, x):
 
         return cls._fft.ifft(x)
 
+    @classmethod
+    def transpose(cls, x):
```
Comment (lostanlen) [inline kymatio/scattering1d/backend/numpy_backend.py:102]: rename to `swap_time_frequency` ?

**Review by lostanlen — CHANGES_REQUESTED** [review]

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 4 above)
Comment (lostanlen) [reply kymatio/scattering1d/backend/numpy_backend.py:102]: i propose renaming because our implementation of `transpose` does a `complex_check`
it also does not exactly behave like `np.transpose`

**Review by rastegah — APPROVED** [review]

**Comment by lostanlen** [comment]
@MuawizChaudhary LMK if/when you need a new review

File: kymatio/scattering1d/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -96,4 +96,12 @@ def ifft(cls, x):
...
         return tf.signal.ifft(x, name='ifft1d')
 
+    @classmethod
+    def swap_time_frequency(cls, x):
+        cls.complex_check(x)
+
+        return tf.linalg.matrix_transpose(x)
+
+
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/backend/tensorflow_backend.py:106]: too much space. delete

**Review by MuawizChaudhary — COMMENTED** [review]
@lostanlen Gopher it

File: tests/scattering1d/test_torch_backend_1d.py (hunk 1 above)
Comment (rastegah) [reply tests/scattering1d/test_torch_backend_1d.py]: fixed

File: tests/scattering1d/test_torch_backend_1d.py (hunk 3 above)
Comment (rastegah) [reply tests/scattering1d/test_torch_backend_1d.py]: fixed

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 4 above)
Comment (rastegah) [reply kymatio/scattering1d/backend/numpy_backend.py]: edited

File: kymatio/scattering1d/backend/torch_backend.py [hunk 6]
```diff
@@ -127,5 +127,10 @@ def ifft(cls, x):
 
         return _ifft(x)
 
+    @classmethod
+    def swap_time_frequency(cls, x):
+        cls.complex_check(x)
+
+        return torch.transpose(x, dim0=-2, dim1=-3)
```
Comment (lostanlen) [inline kymatio/scattering1d/backend/torch_backend.py:134]: maybe add a comment here stating that we expect `axis=-1` to be real/imag parts? hence we swap `-2` vs `-3`

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 7]
```diff
@@ -116,3 +116,26 @@ def coefficent(n):
...
+    shape_T = (10, 20, 5, 3)
+
+    x = np.arange(np.prod(shape)).reshape(shape) * 1j
+    x_T = backend.swap_time_frequency(x)
+    assert tuple(x_T.shape) == shape_T
+
+    x_T_T = backend.swap_time_frequency(x_T)
+    assert tuple(x_T_T.shape) == shape
+    assert x_T_T.shape == x.shape
+    assert np.all(x == x_T_T)
```
Comment (lostanlen) [inline tests/scattering1d/test_tensorflow_backend_1d.py:134]: nice

File: kymatio/scattering1d/backend/torch_backend.py (hunk 6 above)
Comment (rastegah) [reply kymatio/scattering1d/backend/torch_backend.py:146]: bon

**Comment by rastegah** [comment]
> @MuawizChaudhary LMK if/when you need a new review

now

**Review by lostanlen — APPROVED** [review]

**Review by MuawizChaudhary — COMMENTED** [review]
Looks good to me. @lostanlen ?

**Comment by lostanlen** [comment]
@MuawizChaudhary yes, good. nice work

**Comment by MuawizChaudhary** [comment]
@janden can you look over #807, squash and merge, and then look over this PR?

**Comment by lostanlen** [comment]
this is nice but should be merged after v0.3
let's talk about it tomorrow

**Comment by lostanlen** [comment]
yo @cyrusvahidi @MuawizChaudhary now that we have a spin PR #922  i'd like to revive this one
mind if i take this code and make a fresh PR? (with credit)

**Comment by lostanlen** [comment]
superseded by #925 (merged in August 2022)
